etcdmain/etcd_test.go

```diff
@@ -0,0 +1,209 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdmain
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"net/url"
+	"os"
+	"os/exec"
+	"sort"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/coreos/etcd/pkg/transport"
+)
+
+var (
+	testClusterSize = 3
+	electionPause   = 3 * time.Second
+	stopTimeout     = 2 * time.Second
+
+	certPath       = "testcerts/cert.pem"
+	privateKeyPath = "testcerts/key.pem"
+	caPath         = "testcerts/ca.perm"
+	testTLSInfo    = transport.TLSInfo{
+		CertFile: certPath,
+		KeyFile:  privateKeyPath,
+		CAFile:   caPath,
+	}
+)
+
+func Test_startEtcd_v2(t *testing.T) {
+	isClientTLS, isPeerTLS := false, false
+	v2test(t, isClientTLS, isPeerTLS)
+}
+
+func Test_startEtcd_v2_TLS(t *testing.T) {
+	isClientTLS, isPeerTLS := true, true
+	v2test(t, isClientTLS, isPeerTLS)
+}
+
+func Test_startEtcd_v2_TLS_client_to_server(t *testing.T) {
+	isClientTLS, isPeerTLS := true, false
+	v2test(t, isClientTLS, isPeerTLS)
+}
+
+func Test_startEtcd_v2_TLS_server_to_server(t *testing.T) {
+	isClientTLS, isPeerTLS := false, true
+	v2test(t, isClientTLS, isPeerTLS)
+}
+
+// v2test sets up N number of members in a cluster, and tests simple curl client
+// operations and EtcdServer.Stop methods.
+func v2test(t *testing.T, isClientTLS, isPeerTLS bool) {
+	configs, errE := createTestConfigs(testClusterSize, isClientTLS, isPeerTLS)
+	if errE != nil {
+		t.Error(errE)
+	}
+
+	// start the cluster
+	stops := make([]<-chan struct{}, testClusterSize)
+	for i := range configs {
+		cfg := configs[i]
+		defer os.RemoveAll(cfg.dir)
+		stopped, err := startEtcd(cfg)
+		if err != nil {
+			t.Error(err)
+		}
+		stops[i] = stopped
+	}
+
+	time.Sleep(electionPause)
+
+	// curl client endpoint: PUT, GET
+	puts, gets := []string{}, []string{}
+	if !isClientTLS {
+		puts = []string{"curl", "-L", configs[0].acurls[0].String() + "/v2/keys/testKey", "-XPUT", "-d", `value="this is awesome"`}
+		gets = []string{"curl", "-L", configs[0].acurls[0].String() + "/v2/keys/testKey"}
+	} else {
+		// give the CA signed client certificate to the server
+		puts = []string{"curl", "--cacert", caPath, "--cert", certPath, "--key", privateKeyPath, "-L", configs[0].acurls[0].String() + "/v2/keys/testKey", "-XPUT", "-d", `value="this is awesome"`}
+		gets = []string{"curl", "--cacert", caPath, "--cert", certPath, "--key", privateKeyPath, "-L", configs[0].acurls[0].String() + "/v2/keys/testKey"}
+	}
+	if _, err := exec.Command(puts[0], puts[1:]...).Output(); err != nil {
```

Comment:
I have code that runs binaries for testing but mine got complicated when it
deals with process and kills pids.

On Sun, Jan 10, 2016, 17:59 Gyu-Ho Lee <gyuhox@gmail.com> wrote:

> Agree. Binary might be cleaner way to do this? If anybody has better way
> please let me know. I can work that way or close this. Thanks!
>
> On Sun, Jan 10, 2016, 17:53 Xiang Li <notifications@github.com> wrote:
>
>> In etcdmain/etcd_test.go
>> <https://github.com/coreos/etcd/pull/4180#discussion_r49285980>:
>>
>> > +		stops[i] = stopped
>> > +	}
>> > +
>> > +	time.Sleep(electionPause)
>> > +
>> > +	// curl client endpoint: PUT, GET
>> > +	puts, gets := []string{}, []string{}
>> > +	if !isClientTLS {
>> > +		puts = []string{"curl", "-L", configs[0].acurls[0].String() + "/v2/keys/testKey", "-XPUT", "-d", `value="this is awesome"`}
>> > +		gets = []string{"curl", "-L", configs[0].acurls[0].String() + "/v2/keys/testKey"}
>> > +	} else {
>> > +		// give the CA signed client certificate to the server
>> > +		puts = []string{"curl", "--cacert", caPath, "--cert", certPath, "--key", privateKeyPath, "-L", configs[0].acurls[0].String() + "/v2/keys/testKey", "-XPUT", "-d", `value="this is awesome"`}
>> > +		gets = []string{"curl", "--cacert", caPath, "--cert", certPath, "--key", privateKeyPath, "-L", configs[0].acurls[0].String() + "/v2/keys/testKey"}
>> > +	}
>> > +	if _, err := exec.Command(puts[0], puts[1:]...).Output(); err != nil {
>>
>> It would be much easier for us just exec the actual etcd binary than this
>> "unit test", which basically achieves same goal.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/coreos/etcd/pull/4180/files#r49285980>.
>>
> --
> Sincerely,
> Gyu-Ho Lee
>
-- 
Sincerely,
Gyu-Ho Lee